File: funcscript/block/ReferenceBlock.cs
using funcscript.core;

namespace funcscript.block
{
    public class ReferenceBlock : ExpressionBlock
    {
        string _name, _nameLower;
        private bool _fromParent;
        public string Name
        {
            get
            {
                return _name;
            }
            set
            {
                if (value == null)
                {
                    _name = _nameLower = null;
                }
                else
                {
                    _name = value;
                    _nameLower = _name.ToLower();
                }
            }
        }
        public ReferenceBlock(string name)
        {
            Name = name;
            _fromParent = false;
        }
        public ReferenceBlock(string name, string nameLower)
        {
            Name = name;
            _nameLower = nameLower;
            _fromParent = false;
        }
        public ReferenceBlock(string name, string nameLower,bool fromParent)
        {
            Name = name;
            _nameLower = nameLower;
            _fromParent =fromParent;
        }
        public override (object,CodeLocation) Evaluate(IFsDataProvider provider,List<Action> connectionActions)
        {
            if (_fromParent)
                return (provider.ParentProvider?.GetData(_nameLower),this.CodeLocation);
            return (provider.GetData(_nameLower), this.CodeLocation);
        }

        public override IList<ExpressionBlock> GetChilds()
        {
            return new ExpressionBlock[0];
        }
        public override string ToString()
        {
            return Name;
        }

        public override string AsExpString(IFsDataProvider provider)
        {
            return Name;
        }

    }

}

File: js_port/fs_js/src/ast/referenceBlock.js
const { ExpressionBlock } = require('./expressionBlock');
const { literalNull } = require('../core/values');

class ReferenceBlock extends ExpressionBlock {
  constructor(identifier) {
    super();
    this.identifier = identifier;
  }

  evaluate(provider) {
    if (!provider || typeof provider.get !== 'function') {
      return this.makeResult(literalNull());
    }
    const value = provider.get(this.identifier);
    if (value != null) {
      return this.makeResult(value);
    }
    return this.makeResult(literalNull());
  }

  asExpString() {
    return this.identifier;
  }
}

module.exports = {
  ReferenceBlock,
};
